File: presentations/2022_10_25_introduction_to_julia/estimate_pi.py
import numpy as np
import numba
from timeit import default_timer as timer

@numba.njit
def estimate_pi(nMC):
    radius = 1.
    diameter = 2.*radius
    n_circle = 0
    for i in range(nMC):
        x = (np.random.random()-0.5)*diameter
        y = (np.random.random()-0.5)*diameter
        r = np.sqrt(x**2 + y**2)
        if r <= radius:
           n_circle += 1
    return 4.*n_circle/nMC

nMC = 1_000_000

for i in range(3):
    print("function call ", i+1)
    start = timer()
    pi_est = estimate_pi(nMC)
    end = timer()
    print((end - start)*1_000.0, " ms")
    #print(pi_est)

pi_est = estimate_pi(nMC)
print("L1 error: ", np.abs( pi_est - np.pi ) / np.pi)

File: presentations/2022_10_25_introduction_to_julia/estimate_pi.jl
using Pkg
Pkg.activate(".")
Pkg.instantiate()

@info "loading packages"
@time using BenchmarkTools
using Random
@info "done"

function estimate_pi(nMC)

    radius   = 1.0
    diameter = 2radius
    n_circle = 0
    
    @inbounds for _ ∈ 1:nMC
        x = (rand() - 0.5) * diameter
        y = (rand() - 0.5) * diameter
        r = √(x^2 + y^2)
        if r <= radius
           n_circle += 1
        end
    end
    
    return 4n_circle/nMC
end

nMC = 1_000_000


@info "First function call"
@time estimate_pi(nMC)

@info "Second function call"
@time estimate_pi(nMC)

@info "benchmark:"
@btime estimate_pi(nMC)

@info "L1 error: $( abs(estimate_pi(nMC) - π) / π )"
